### PR TITLE
Oceanwater 775 - Fault Handling Pattern 6

### PR DESCRIPTION
**Pull request by scheidere**
This pattern is difficult to review beyond a code review, and the same testing we did for Pattern 5. The additions to it are rooted in battery level and power fault handling, which both have issues or limitations in injection capabilities.

### CODE REVIEW

**Comment by kmdalal** [comment]
I had long forgotten that Pattern 6 was to use power faults, which are not easy to model and may not be working correctly or as desirable in the simulator.  I'm not sure how much of this pattern can work in this state, but I'm having a look.

File: ow_plexil/src/plans/FaultHandlingPattern6.plp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+LibraryAction MonitorPower (In Boolean continue,
+                            InOut Boolean battery_temp_ok,
+                            InOut Boolean battery_life_ok,
+                            InOut Boolean battery_charge_ok,
+                            InOut Boolean all_ok);
+
+LibraryAction MissionSequence2 (InOut Boolean mission_in_progress,
+                       InOut Boolean pan_done,
+                       InOut Boolean unstow_done,
+                       InOut Boolean stow_done);
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern6.plp:47]: These and many more lines below need re-indentation.

File: ow_plexil/src/plans/FaultHandlingPattern6.plp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  RunMission: Concurrence
+  {
+    
+    End !MissionInProgress;
+    log_info ("Starting Mission...");
+
+    LogPanError:
+    {
+
+      Start !PanDone && (Lookup(AntennaFault) || Lookup(PowerFault) || (Lookup(StateOfCharge) < CriticallyLowCharge));
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern6.plp:77]: Line long, hard to read.

File: ow_plexil/src/plans/FaultHandlingPattern6.plp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                            battery_charge_ok = BatteryChargeOK,
+                            all_ok = BatteryOK);
+
+
+    BackupPanning:
+    {
+      Real NewAngle;
+
+      Repeat MissionInProgress;
+      Start MissionInProgress && Lookup(ArmFault) && PanDone && (Lookup(StateOfCharge) > CriticallyLowCharge) && !Lookup(PowerFault);
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern6.plp:153]: Line too long.

File: ow_plexil/src/plans/FaultHandlingPattern6.plp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        log_error ("Panning halted due to the battery level being critically low.");
+      }
+      endif;
+
+    }
+
+    LogUnstowError:
+    {
+
+      Start PanDone && !UnstowDone && (Lookup(ArmFault) || Lookup(PowerFault) || !BatteryOK);
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern6.plp:99]: Although operator precedence yields the desired logic here, the code would be clearer if the fault disjunction was put in parentheses.

File: ow_plexil/src/plans/FaultHandlingPattern6.plp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        log_error ("Unstowing halted due to the battery level being low.");
+      }
+      endif;
+
+    }
+
+    LogStowError:
+    {
+
+      Start UnstowDone && !StowDone && (Lookup(ArmFault) || Lookup(PowerFault) || !BatteryOK);
```
Comment (kmdalal) [inline ow_plexil/src/plans/FaultHandlingPattern6.plp:121]: Same as previous comment.

File: ow_plexil/src/plans/FaultHandlingPattern6.plp (hunk 4 above)
Comment (scheidere) [reply ow_plexil/src/plans/FaultHandlingPattern6.plp:99]: Do you mean around the lookups for arm and power faults?

Comment (kmdalal) [reply ow_plexil/src/plans/FaultHandlingPattern6.plp:99]: and !BatteryOK

Comment (scheidere) [reply ow_plexil/src/plans/FaultHandlingPattern6.plp:99]: I think I am missing something. There are parentheses around those three terms as is right? What am I adding?

Comment (kmdalal) [reply ow_plexil/src/plans/FaultHandlingPattern6.plp:100]: Indeed they are already bracketed.   I just didn't see it the first time.  Sorry for the confusion!

**Comment by kmdalal** [comment]
Runs successfully without injecting any faults.   When I tried injecting arm and antenna faults, I was surprised that they weren't detected at all -- in the simulator or autonomy.   It's as if I were using the latest melodic-devel of ow_simulator, but I'm still on the older branch.

**Comment by scheidere** [comment]
That is very odd, I just tested it again, and both antenna and arm faults are detected and halt actions as we expect. The backup panning works as well.

**Comment by kmdalal** [comment]
This branch needs melodic-devel merged into it.

**Comment by scheidere** [comment]
I did a clean rebuild and both fault types still work for me.

**Comment by kmdalal** [comment]
Even though my ow_simulator showed to be on 6feae0d, it took several steps and two rebuilds to get things working as expected again.

**Review by kmdalal — APPROVED** [review]
The pattern looks reasonable and correct as coded, though I can only test it as far as antenna and arm faults impact it, as power fault injection is essentially unusable in the simulator.   To this end, this plan behaves the same as pattern 5.   I think it's fine to merge, and we can retry it when power fault injection is improved.   Nice work!

**Pull request merged**